pkg/operator/cockroachdb/controller.go

```diff
@@ -0,0 +1,573 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/kubelet/apis"
+)
+
+const (
+	CustomResourceName             = "cluster"
+	CustomResourceNamePlural       = "clusters"
+	appName                        = "rook-cockroachdb"
+	createInitRetryIntervalDefault = 6 * time.Second
+	createInitTimeout              = 1 * time.Minute
+	updateClusterInterval          = 30 * time.Second
+	updateClusterTimeout           = 1 * time.Hour
+	httpPortDefault                = int32(8080)
+	httpPortName                   = "http"
+	grpcPortDefault                = int32(26257)
+	grpcPortName                   = "grpc"
+	volumeNameDataDir              = "datadir"
+	envVarChannel                  = "COCKROACH_CHANNEL"
+	envVarValChannelSecure         = "kubernetes-secure"
+	envVarValChannelInsecure       = "kubernetes-insecure"
+)
+
+var ClusterResource = opkit.CustomResource{
+	Name:    CustomResourceName,
+	Plural:  CustomResourceNamePlural,
+	Group:   cockroachdbv1alpha1.CustomResourceGroup,
+	Version: cockroachdbv1alpha1.Version,
+	Scope:   apiextensionsv1beta1.NamespaceScoped,
+	Kind:    reflect.TypeOf(cockroachdbv1alpha1.Cluster{}).Name(),
+}
+
+type ClusterController struct {
+	context                 *clusterd.Context
+	containerImage          string
+	createInitRetryInterval time.Duration
+}
+
+func NewClusterController(context *clusterd.Context, containerImage string) *ClusterController {
+	return &ClusterController{
+		context:                 context,
+		containerImage:          containerImage,
+		createInitRetryInterval: createInitRetryIntervalDefault,
+	}
+}
+
+type cluster struct {
+	context   *clusterd.Context
+	namespace string
+	spec      cockroachdbv1alpha1.ClusterSpec
+	ownerRef  metav1.OwnerReference
+}
+
+func newCluster(c *cockroachdbv1alpha1.Cluster, context *clusterd.Context) *cluster {
+	return &cluster{
+		context:   context,
+		namespace: c.Namespace,
+		spec:      c.Spec,
+		ownerRef:  clusterOwnerRef(c.Namespace, string(c.UID)),
+	}
+}
+
+func clusterOwnerRef(namespace, clusterID string) metav1.OwnerReference {
+	blockOwner := true
+	return metav1.OwnerReference{
+		APIVersion:         ClusterResource.Version,
+		Kind:               ClusterResource.Kind,
+		Name:               namespace,
+		UID:                types.UID(clusterID),
+		BlockOwnerDeletion: &blockOwner,
+	}
+}
+
+func (c *ClusterController) StartWatch(namespace string, stopCh chan struct{}) error {
+
+	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	}
+
+	logger.Infof("start watching cockroachdb clusters in all namespaces")
+	watcher := opkit.NewWatcher(ClusterResource, namespace, resourceHandlerFuncs, c.context.RookClientset.CockroachdbV1alpha1().RESTClient())
+	go watcher.Watch(&cockroachdbv1alpha1.Cluster{}, stopCh)
+
+	return nil
+}
+
+func (c *ClusterController) onAdd(obj interface{}) {
+	clusterObj := obj.(*cockroachdbv1alpha1.Cluster)
```

Comment:
i think you want to call `DeepCopy()` after you cast it

Reply:
+1

Reply:
Good catch on the missing `DeepCopy()` guys, thanks!